In [1]:
# Importing Required Packages
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
pd.set_option('display.max_columns', None)# To see all the columns of a dataframe
#pd.set_option('display.max_rows', None)

In [2]:
# Loadiing reduced feature training set
X_train = pd.read_csv('X_train_final.csv')
y_train = pd.read_csv('y_train.final.csv')

In [3]:
# Loadiing reduced feature test set
X_test = pd.read_csv('X_test_final.csv')
y_test = pd.read_csv('y_test.final.csv')

## Model_16: Weighted Aggregating Classifier 

In [4]:
# Importing Joblib module
import joblib

In [5]:
import tensorflow as tf
from tensorflow import keras

In [6]:
# Importing best Voting Classifier i.e. voting Classifier_3
voting_clf = joblib.load('Voting_class_3.joblib')

In [7]:
# Importing best Neural Net Classifier i.e. Dense Neural Net with RELU activation and Nadam optimizer
neural_clf = keras.models.load_model('Best_model_Elu_eq_Learn_relu.h5')

### Performing 5 fold class validations to get best weights for Voting and Neural Calssifiers 

In [8]:
from sklearn.model_selection import StratifiedKFold

In [9]:
# Initializing the 5 Fold  Stratified fold object
cv_strat = StratifiedKFold(5,random_state=42)

In [10]:
# Importing the Sklearn's roc_auc_score module
from sklearn.metrics import roc_auc_score

In [11]:
# Importing cross val score from sklearn
from sklearn.model_selection import cross_val_score

In [12]:
# Importing train test split from Sklearn to produce validation set
from sklearn.model_selection import train_test_split

In [13]:
# Instantiating an empty DataFrame to store predictions of the Voting & Neural classifier on the test fold
df_pred = pd.DataFrame(columns=['Voting','Neural','True'])

In [14]:
# Defining the exponential decay learning rate with tuned step parameter = 2 for neural classifier with RELU activation
def exponential_decay_fn(epoch):
            return 0.01 * 0.1**(epoch /2 )

In [15]:
def Neural_Pred(clf, X_tr, y_tr, X_tst):
    ''' 
    This function returns the predictions of Neural Classifier(clf) for each of the test fold. 
    Parameters:
    X_tr : Training set Features
    y_tr : Training set Labels
    X_tst : Test set Features
   
    '''
    # Splitting the Training set further into training & validation set.
    X_tr_r, X_val, y_tr_r, y_val = train_test_split(X_tr, y_tr, test_size=0.15, random_state=42, stratify=y_tr)
    
    # Compiling & Training the Neural Net
    clf.compile(loss="binary_crossentropy", optimizer=keras.optimizers.Nadam(beta_1=0.9, beta_2=0.999),
                metrics=[tf.keras.metrics.AUC(curve="ROC")])
    
    # defining Checkpoints
    checkpoint_cb = keras.callbacks.ModelCheckpoint('best_model.h5',save_best_only=True) # 1st Callback
    early_stopping_cb = keras.callbacks.EarlyStopping(patience=4,restore_best_weights=True)# 2nd Callback
    lr_scheduler_cb = keras.callbacks.LearningRateScheduler(exponential_decay_fn)# 3rd Callback
    
    # Fitting The model
    clf.fit(X_tr_r, y_tr_r, epochs=50, validation_data=(X_val,y_val),batch_size=32,
    class_weight={0: 1.0, 1: 12.0},callbacks=[checkpoint_cb,early_stopping_cb,lr_scheduler_cb])
    
    # Loading the best Neural model after training & before making predictions
    clf_best = keras.models.load_model('best_model.h5')    
    
    # returning the predictions for the test fold.
    return (clf_best.predict_proba(X_tst).flatten())

In [16]:
# Computing Voting Classifier's & Neural Net predictions on the test fold
for train_index, test_index in cv_strat.split(X_train, y_train):
    # Creating  Folds
    X_tr, X_tst = X_train.loc[train_index], X_train.loc[test_index]
    y_tr, y_tst = y_train.loc[train_index], y_train.loc[test_index]
    
    # Fitting the best Voting Classifier on the Training Folds
    voting_clf.fit(X_tr.values, y_tr)
    # Making Predictions on the testing Fold.
    y_pred_voting = voting_clf.predict_proba(X_tst.values)[:,1]
    
    
    # Loading the fresh best Neural net classifier
    neural = keras.models.load_model('Best_model_Elu_eq_Learn_relu.h5')
    # Fitting the best Neural Net on Training Folds & obtaining  predictions
    y_pred_neural = Neural_Pred(neural, X_tr, y_tr, X_tst)
    
    # Storing the predictions in a dataframe
    df_temp = pd.DataFrame(columns=['Voting','Neural','True'])
    df_temp['Voting'] = y_pred_voting
    df_temp['Neural'] = y_pred_neural
    df_temp['True'] = y_tst.values.flatten() # As y_tst is a dataframe we first convert it into a 2d numpy vector
    # array, which is is then converted to 1d array using .flatten()
    
    # Appending the df_temp to df_train_stack
    df_pred = df_pred.append(df_temp,ignore_index=True)   
    

  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 156830 samples, validate on 27676 samples
Epoch 1/50
156830/156830 [==============================] - 24s 150us/sample - loss: 1.1223 - auc: 0.7484 - val_loss: 1.4785 - val_auc: 0.7539
Epoch 2/50
156830/156830 [==============================] - 22s 139us/sample - loss: 1.1015 - auc: 0.7598 - val_loss: 4.0262 - val_auc: 0.7607
Epoch 3/50
156830/156830 [==============================] - 22s 140us/sample - loss: 1.0917 - auc: 0.7648 - val_loss: 2.3824 - val_auc: 0.7563
Epoch 4/50
156830/156830 [==============================] - 22s 142us/sample - loss: 1.0849 - auc: 0.7688 - val_loss: 4.3395 - val_auc: 0.7558
Epoch 5/50
156830/156830 [==============================] - 22s 141us/sample - loss: 1.0815 - auc: 0.7705 - val_loss: 3.5095 - val_auc: 0.7565
  ...
    to  
  ['...']
  ...
    to  
  ['...']
Train on 156830 samples, validate on 27676 samples
Epoch 1/50
156830/156830 [==============================] - 25s 158us/sample - loss

In [17]:
# Visualizing the head of the df_pred
df_pred.head()

Voting    Neural True
0  0.453474  0.583205    0
1  0.114683  0.179734    0
2  0.109974  0.177286    0
3  0.472716  0.512960    1
4  0.439749  0.525648    1

In [19]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230633 entries, 0 to 230632
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Voting  230633 non-null  float64
 1   Neural  230633 non-null  float32
 2   True    230633 non-null  object 
dtypes: float32(1), float64(1), object(1)
memory usage: 4.4+ MB


In [21]:
# Converting the True column to int
df_pred['True'] = df_pred["True"].astype(np.int)

In [22]:
df_pred.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230633 entries, 0 to 230632
Data columns (total 3 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Voting  230633 non-null  float64
 1   Neural  230633 non-null  float32
 2   True    230633 non-null  int32  
dtypes: float32(1), float64(1), int32(1)
memory usage: 3.5 MB


In [23]:
# importing hyper-parameter tuner optuna
import optuna

In [24]:
# Importing the Sklearn's roc_auc_score module
from sklearn.metrics import roc_auc_score

In [25]:
# Declaring objective function for Weighted Aggregation Classifier
def objective_wrappper_auc(y_pred_cls_1, y_pred_cls_2, y_true):
    '''
    Optimizes weights for predictions from 2 classifiers to achieve maximum roc_auc score
      
    '''
    
    def objective(trial):
        w1 = trial.suggest_uniform('w1',-1,1)
        w2 = trial.suggest_uniform('w2',-1,1)
        
        y_final_pred = y_pred_cls_1 * w1 + y_pred_cls_2 * w2
        
        roc_auc_1 = roc_auc_score(y_true,y_final_pred)
        
                   
        return roc_auc_1
    
    return objective

In [29]:
# Defining the evaluation function for study's best parameters
def roc_auc_weights(y_pred_cls_1, y_pred_cls_2, y_true, obj_func, n_trials=200):
    ''' Computes the best hyper parameters of the classsifier on the Training set and returns 
    Optuna's study's best score & clasifier parameters'''
    study = optuna.create_study(direction='maximize')
    study.optimize(obj_func(y_pred_cls_1, y_pred_cls_2, y_true),n_trials)
    best_score = study.best_value
    best_params = study.best_params
    return (best_score,best_params)


In [31]:
best_study_score,best_study_params = roc_auc_weights(df_pred['Voting'].values, df_pred['Neural'].values, df_pred['True'].values, 
                                                     objective_wrappper_auc,n_trials=2000)

[I 2021-01-09 02:44:06,427] A new study created in memory with name: no-name-220e5526-1422-4220-b024-fc875062bcb3
[I 2021-01-09 02:44:06,485] Trial 0 finished with value: 0.7620881569365854 and parameters: {'w1': 0.22822449452402727, 'w2': 0.5357813070274884}. Best is trial 0 with value: 0.7620881569365854.
[I 2021-01-09 02:44:06,537] Trial 1 finished with value: 0.6123382863184641 and parameters: {'w1': 0.47704777513389085, 'w2': -0.41782190575282896}. Best is trial 0 with value: 0.7620881569365854.
[I 2021-01-09 02:44:06,589] Trial 2 finished with value: 0.4633859620054448 and parameters: {'w1': 0.8857141337655585, 'w2': -0.947698904818205}. Best is trial 0 with value: 0.7620881569365854.
[I 2021-01-09 02:44:06,639] Trial 3 finished with value: 0.7496537260482155 and parameters: {'w1': -0.04835498516090819, 'w2': 0.35807712330755703}. Best is trial 0 with value: 0.7620881569365854.
[I 2021-01-09 02:44:06,691] Trial 4 finished with value: 0.23610975631867714 and parameters: {'w1': -0.

[I 2021-01-09 02:44:08,660] Trial 39 finished with value: 0.7684869697356086 and parameters: {'w1': 0.8465925379266503, 'w2': 0.26868551372224525}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:08,717] Trial 40 finished with value: 0.7686708461343559 and parameters: {'w1': 0.9908457367024549, 'w2': 0.043478086221006865}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:08,774] Trial 41 finished with value: 0.7687699435689779 and parameters: {'w1': 0.7420479034998689, 'w2': 0.13391923617305265}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:08,831] Trial 42 finished with value: 0.7687692694695124 and parameters: {'w1': 0.7379560806034049, 'w2': 0.13373931423184213}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:08,888] Trial 43 finished with value: 0.7683968828398404 and parameters: {'w1': 0.5608719425752431, 'w2': -0.010482100635903158}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-0

[I 2021-01-09 02:44:10,955] Trial 79 finished with value: 0.766008008343196 and parameters: {'w1': 0.3137397521234575, 'w2': 0.3017455659148917}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:11,009] Trial 80 finished with value: 0.7687843922488414 and parameters: {'w1': 0.6159484079869095, 'w2': 0.1003016269325396}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:11,064] Trial 81 finished with value: 0.7687911081642609 and parameters: {'w1': 0.587856180811245, 'w2': 0.08826770557515581}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:11,119] Trial 82 finished with value: 0.7686837117824419 and parameters: {'w1': 0.6987537737939231, 'w2': 0.1658866196220564}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:11,174] Trial 83 finished with value: 0.768097223208088 and parameters: {'w1': 0.5445774356997644, 'w2': 0.23678853002858685}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:1

[I 2021-01-09 02:44:13,156] Trial 118 finished with value: 0.768338320037117 and parameters: {'w1': 0.5243459630546169, 'w2': -0.014687923673378483}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:13,211] Trial 119 finished with value: 0.7674468674459267 and parameters: {'w1': 0.5932926731892919, 'w2': -0.07160462123085357}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:13,266] Trial 120 finished with value: 0.7687725936082523 and parameters: {'w1': 0.6643813136269865, 'w2': 0.0630824677007488}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:13,321] Trial 121 finished with value: 0.7687830807821248 and parameters: {'w1': 0.7669538135957695, 'w2': 0.12631376643597786}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:13,376] Trial 122 finished with value: 0.7686580643370491 and parameters: {'w1': 0.7189715382736738, 'w2': 0.028425599485914096}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-

[I 2021-01-09 02:44:15,342] Trial 157 finished with value: 0.7686342256365581 and parameters: {'w1': 0.6976288136320786, 'w2': 0.18166905373161368}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:15,397] Trial 158 finished with value: 0.7681079262153858 and parameters: {'w1': 0.7786998685881652, 'w2': -0.04604467889564315}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:15,452] Trial 159 finished with value: 0.7687913885957183 and parameters: {'w1': 0.734106503444008, 'w2': 0.1093865177034191}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:15,510] Trial 160 finished with value: 0.7686969870580498 and parameters: {'w1': 0.9608106537349438, 'w2': 0.22163578388541044}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-09 02:44:15,566] Trial 161 finished with value: 0.768792947814888 and parameters: {'w1': 0.8059286970139098, 'w2': 0.11645125266647632}. Best is trial 31 with value: 0.7687940520612504.
[I 2021-01-0

[I 2021-01-09 02:44:17,549] Trial 196 finished with value: 0.23392118056052194 and parameters: {'w1': -0.9015525779880825, 'w2': 0.17977071490239882}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:17,605] Trial 197 finished with value: 0.76835494630398 and parameters: {'w1': 0.8062530543572091, 'w2': -0.02053823263875673}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:17,661] Trial 198 finished with value: 0.7687484357174846 and parameters: {'w1': 0.9291955684547355, 'w2': 0.07226344504194454}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:17,718] Trial 199 finished with value: 0.7687368650699502 and parameters: {'w1': 0.6855906135050162, 'w2': 0.14209932884063006}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:17,775] Trial 200 finished with value: 0.7687857346212834 and parameters: {'w1': 0.8910727402598663, 'w2': 0.0977224249944682}. Best is trial 181 with value: 0.7687940713139974.
[I 20

[I 2021-01-09 02:44:19,863] Trial 235 finished with value: 0.7687656370334972 and parameters: {'w1': 0.6322711077005326, 'w2': 0.11692906307877142}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:19,944] Trial 236 finished with value: 0.7687094121728886 and parameters: {'w1': 0.836153020685353, 'w2': 0.187126843540973}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:20,004] Trial 237 finished with value: 0.7687356278276393 and parameters: {'w1': 0.7697643862568864, 'w2': 0.0543748222870295}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:20,060] Trial 238 finished with value: 0.7687936247009355 and parameters: {'w1': 0.8875297065093869, 'w2': 0.11540058654439492}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-01-09 02:44:20,116] Trial 239 finished with value: 0.7686068150448744 and parameters: {'w1': 0.9021865643310042, 'w2': 0.022207466611086787}. Best is trial 181 with value: 0.7687940713139974.
[I 2021-

[I 2021-01-09 02:44:22,274] Trial 274 finished with value: 0.7686056991455337 and parameters: {'w1': 0.9015000219347777, 'w2': 0.021939943504516835}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:22,338] Trial 275 finished with value: 0.7687863785750005 and parameters: {'w1': 0.7762106838915078, 'w2': 0.08582092685669361}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:22,402] Trial 276 finished with value: 0.7687778126226037 and parameters: {'w1': 0.8414120670335282, 'w2': 0.14450493318486024}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:22,461] Trial 277 finished with value: 0.7679700742679726 and parameters: {'w1': 0.11059364238042886, 'w2': 0.05187575599122587}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:22,518] Trial 278 finished with value: 0.7687927697269796 and parameters: {'w1': 0.7059060320742531, 'w2': 0.10238819602641144}. Best is trial 260 with value: 0.7687941288189122.
[I 

[I 2021-01-09 02:44:24,528] Trial 313 finished with value: 0.7687939616240054 and parameters: {'w1': 0.8211611613896804, 'w2': 0.1128102818586245}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:24,584] Trial 314 finished with value: 0.7684625704255283 and parameters: {'w1': 0.9089221854198961, 'w2': -0.0063894693065380676}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:24,644] Trial 315 finished with value: 0.7687851747716709 and parameters: {'w1': 0.8583815271383035, 'w2': 0.09311819906644428}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:24,701] Trial 316 finished with value: 0.7687430408951553 and parameters: {'w1': 0.7596203583383963, 'w2': 0.15421952109661136}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:24,759] Trial 317 finished with value: 0.7687465877071121 and parameters: {'w1': 0.8333070717310829, 'w2': 0.06386268268917815}. Best is trial 260 with value: 0.7687941288189122.
[I 

[I 2021-01-09 02:44:26,808] Trial 352 finished with value: 0.7687574480296178 and parameters: {'w1': 0.9740060428510782, 'w2': 0.1875347006421409}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:26,866] Trial 353 finished with value: 0.7687927074088781 and parameters: {'w1': 0.993391716603436, 'w2': 0.1450958470668024}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:26,934] Trial 354 finished with value: 0.7686565279171849 and parameters: {'w1': 0.9108091728935791, 'w2': 0.035551098075865775}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:26,992] Trial 355 finished with value: 0.7685296936074851 and parameters: {'w1': 0.7382419776210362, 'w2': 0.22295740160795602}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:27,050] Trial 356 finished with value: 0.768793837747779 and parameters: {'w1': 0.6412305993532716, 'w2': 0.08454702359131634}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-

[I 2021-01-09 02:44:29,084] Trial 391 finished with value: 0.768792678276432 and parameters: {'w1': 0.7557776967416886, 'w2': 0.09476524838141345}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:29,143] Trial 392 finished with value: 0.7678001208984345 and parameters: {'w1': 0.677441067205964, 'w2': -0.061779320508398794}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:29,198] Trial 393 finished with value: 0.7686724995906549 and parameters: {'w1': 0.7878947829672064, 'w2': 0.03506041913860648}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:29,255] Trial 394 finished with value: 0.7687206704699376 and parameters: {'w1': 0.7225443869956999, 'w2': 0.1571367614346323}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:29,313] Trial 395 finished with value: 0.7687856201181048 and parameters: {'w1': 0.8151953326417669, 'w2': 0.08923458421577436}. Best is trial 260 with value: 0.7687941288189122.
[I 202

[I 2021-01-09 02:44:31,430] Trial 430 finished with value: 0.7684706051541075 and parameters: {'w1': 0.7259698962023151, 'w2': 0.2343213333123479}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:31,488] Trial 431 finished with value: 0.7687941164059571 and parameters: {'w1': 0.7748756025945158, 'w2': 0.10486028615711969}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:31,547] Trial 432 finished with value: 0.2339847070253754 and parameters: {'w1': -0.37875539815933, 'w2': 0.07660793998309191}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:31,606] Trial 433 finished with value: 0.7687898058172664 and parameters: {'w1': 0.7693923006083443, 'w2': 0.11791774737034544}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:31,664] Trial 434 finished with value: 0.7687374672249407 and parameters: {'w1': 0.6457523413690766, 'w2': 0.046148464846881274}. Best is trial 260 with value: 0.7687941288189122.
[I 202

[I 2021-01-09 02:44:33,733] Trial 469 finished with value: 0.7687927525008377 and parameters: {'w1': 0.8755098923263778, 'w2': 0.12675294091503256}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:33,801] Trial 470 finished with value: 0.7684657321825481 and parameters: {'w1': 0.7011727822754705, 'w2': -0.004527324656866616}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:33,868] Trial 471 finished with value: 0.7686594021496298 and parameters: {'w1': 0.8256588286929206, 'w2': 0.20564806621853787}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:33,944] Trial 472 finished with value: 0.7687574989480667 and parameters: {'w1': 0.5483526071241991, 'w2': 0.1055461121789236}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:34,004] Trial 473 finished with value: 0.7687335145853462 and parameters: {'w1': 0.945845504699243, 'w2': 0.06579095230576826}. Best is trial 260 with value: 0.7687941288189122.
[I 20

[I 2021-01-09 02:44:36,102] Trial 508 finished with value: 0.7687532722101551 and parameters: {'w1': 0.9011305628554209, 'w2': 0.07287753076817526}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:36,160] Trial 509 finished with value: 0.7687348906500951 and parameters: {'w1': 0.8669694014308957, 'w2': 0.1809389572801352}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:36,217] Trial 510 finished with value: 0.768787598337844 and parameters: {'w1': 0.9622060496517137, 'w2': 0.10817703063986239}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:36,276] Trial 511 finished with value: 0.7685843209967559 and parameters: {'w1': 0.8133794632651776, 'w2': 0.015353322896255489}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:36,335] Trial 512 finished with value: 0.7687641890242732 and parameters: {'w1': 0.7871574219246636, 'w2': 0.0693968955947474}. Best is trial 260 with value: 0.7687941288189122.
[I 2021

[I 2021-01-09 02:44:38,397] Trial 547 finished with value: 0.768746882324804 and parameters: {'w1': 0.9298954017936906, 'w2': 0.18624263619286285}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:38,455] Trial 548 finished with value: 0.7687915291914356 and parameters: {'w1': 0.9109275890830549, 'w2': 0.1355264179099642}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:38,513] Trial 549 finished with value: 0.7687801492980905 and parameters: {'w1': 0.9922958218328669, 'w2': 0.10158712687767536}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:38,572] Trial 550 finished with value: 0.7681238553301523 and parameters: {'w1': 0.87827091375166, 'w2': -0.05026640434216429}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:38,630] Trial 551 finished with value: 0.768779900532335 and parameters: {'w1': 0.9789249508361646, 'w2': 0.16543597210013739}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-0

[I 2021-01-09 02:44:40,698] Trial 586 finished with value: 0.7685037502777722 and parameters: {'w1': 0.7734718200720181, 'w2': 0.240780073346963}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:40,758] Trial 587 finished with value: 0.7687729211582744 and parameters: {'w1': 0.6915177509672933, 'w2': 0.12269163160313865}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:40,817] Trial 588 finished with value: 0.7683294898661022 and parameters: {'w1': 0.8438617778768894, 'w2': -0.024757653512104932}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:40,878] Trial 589 finished with value: 0.7687266053824814 and parameters: {'w1': 0.7484332704136593, 'w2': 0.16002070866644771}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:40,941] Trial 590 finished with value: 0.768698808722558 and parameters: {'w1': 0.8073477204771834, 'w2': 0.04378073639119516}. Best is trial 260 with value: 0.7687941288189122.
[I 202

[I 2021-01-09 02:44:43,026] Trial 625 finished with value: 0.7687783253536515 and parameters: {'w1': 0.6843535017768576, 'w2': 0.11705157079692263}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:43,084] Trial 626 finished with value: 0.7687206560303775 and parameters: {'w1': 0.8467502897402169, 'w2': 0.053625793978290925}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:43,142] Trial 627 finished with value: 0.7687724671987697 and parameters: {'w1': 0.7567571340615621, 'w2': 0.13475086320730234}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:43,200] Trial 628 finished with value: 0.7687335829832627 and parameters: {'w1': 0.8870785918597913, 'w2': 0.18580892488640166}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:43,258] Trial 629 finished with value: 0.7687748943115015 and parameters: {'w1': 0.8101442892477329, 'w2': 0.07872062958095857}. Best is trial 260 with value: 0.7687941288189122.
[I 2

[I 2021-01-09 02:44:45,340] Trial 664 finished with value: 0.7685713960704758 and parameters: {'w1': 0.713546195793802, 'w2': 0.20425759497893442}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:45,399] Trial 665 finished with value: 0.7687723200165865 and parameters: {'w1': 0.8373798872633256, 'w2': 0.14927958023900145}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:45,456] Trial 666 finished with value: 0.7687286740394659 and parameters: {'w1': 0.7731372355756206, 'w2': 0.05190903408510104}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:45,514] Trial 667 finished with value: 0.7687921376795664 and parameters: {'w1': 0.8598853743808226, 'w2': 0.10579066426016909}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:45,573] Trial 668 finished with value: 0.7687150111756649 and parameters: {'w1': 0.8054357503248649, 'w2': 0.1777251610456586}. Best is trial 260 with value: 0.7687941288189122.
[I 2021

[I 2021-01-09 02:44:47,666] Trial 703 finished with value: 0.7687325884268923 and parameters: {'w1': 0.6391445634358177, 'w2': 0.13428130671487573}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:47,725] Trial 704 finished with value: 0.7687853125808063 and parameters: {'w1': 0.7405869703707042, 'w2': 0.0805317144545635}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:47,784] Trial 705 finished with value: 0.7687429220854411 and parameters: {'w1': 0.8007928531451084, 'w2': 0.1626471201795065}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:47,844] Trial 706 finished with value: 0.76879012576752 and parameters: {'w1': 0.6663923074821679, 'w2': 0.10186135445843549}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:47,904] Trial 707 finished with value: 0.7687371688073645 and parameters: {'w1': 0.7280879088403024, 'w2': 0.05194380983634632}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-0

[I 2021-01-09 02:44:50,012] Trial 742 finished with value: 0.2324320834414778 and parameters: {'w1': -0.9847244974334068, 'w2': 0.10952077760649347}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:50,073] Trial 743 finished with value: 0.7687606308126637 and parameters: {'w1': 0.7775348422561351, 'w2': 0.0664983063311288}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:50,132] Trial 744 finished with value: 0.768787550205977 and parameters: {'w1': 0.9140566190650932, 'w2': 0.1442026862421658}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:50,190] Trial 745 finished with value: 0.7687937351509042 and parameters: {'w1': 0.7474572272650319, 'w2': 0.10448199451517017}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:50,248] Trial 746 finished with value: 0.7687357473973302 and parameters: {'w1': 0.8647021638158318, 'w2': 0.17993889037836822}. Best is trial 260 with value: 0.7687941288189122.
[I 2021

[I 2021-01-09 02:44:52,355] Trial 781 finished with value: 0.257848685106269 and parameters: {'w1': -0.4116520144752944, 'w2': 0.22322678926004813}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:52,414] Trial 782 finished with value: 0.7687905305818552 and parameters: {'w1': 0.9259656552897297, 'w2': 0.14068327150019233}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:52,474] Trial 783 finished with value: 0.7687627823071248 and parameters: {'w1': 0.7434137324400828, 'w2': 0.06471480991391852}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:52,533] Trial 784 finished with value: 0.7687895611047209 and parameters: {'w1': 0.8562172541886164, 'w2': 0.09905030777164665}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:52,593] Trial 785 finished with value: 0.768758710857816 and parameters: {'w1': 0.7895098379722453, 'w2': 0.15110292299401792}. Best is trial 260 with value: 0.7687941288189122.
[I 202

[I 2021-01-09 02:44:54,702] Trial 820 finished with value: 0.7687746906377058 and parameters: {'w1': 0.7567177124606161, 'w2': 0.0733331355846154}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:54,761] Trial 821 finished with value: 0.7673370839768233 and parameters: {'w1': 0.8370575623997464, 'w2': -0.10776076354066512}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:54,821] Trial 822 finished with value: 0.7686820170340675 and parameters: {'w1': 0.7342768400179304, 'w2': 0.035149414620019345}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:54,882] Trial 823 finished with value: 0.7686648699297367 and parameters: {'w1': 0.900996855450471, 'w2': 0.22206304498357263}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:54,943] Trial 824 finished with value: 0.7687916097489815 and parameters: {'w1': 0.805446001166711, 'w2': 0.11952358431637589}. Best is trial 260 with value: 0.7687941288189122.
[I 202

[I 2021-01-09 02:44:57,072] Trial 859 finished with value: 0.7687883084095473 and parameters: {'w1': 0.7290332546771794, 'w2': 0.11385767541771699}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:57,135] Trial 860 finished with value: 0.7686501218189945 and parameters: {'w1': 0.8633541592158173, 'w2': 0.031963239597242094}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:57,194] Trial 861 finished with value: 0.7687588755194669 and parameters: {'w1': 0.7782188868286571, 'w2': 0.14885943362963908}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:57,254] Trial 862 finished with value: 0.7683364223749236 and parameters: {'w1': 0.6577751833836134, 'w2': -0.01861496873326219}. Best is trial 260 with value: 0.7687941288189122.
[I 2021-01-09 02:44:57,313] Trial 863 finished with value: 0.7687727534567163 and parameters: {'w1': 0.8164245414826344, 'w2': 0.0776664886638267}. Best is trial 260 with value: 0.7687941288189122.
[I 2

[I 2021-01-09 02:44:59,436] Trial 898 finished with value: 0.7687284989914649 and parameters: {'w1': 0.726439056116098, 'w2': 0.1544906464132504}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:44:59,495] Trial 899 finished with value: 0.23344850485252794 and parameters: {'w1': -0.06107408664231284, 'w2': 0.010775565886005561}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:44:59,555] Trial 900 finished with value: 0.7687753421911915 and parameters: {'w1': 0.8144140627664471, 'w2': 0.07951223859227163}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:44:59,615] Trial 901 finished with value: 0.7687937744163746 and parameters: {'w1': 0.8826844360223511, 'w2': 0.11614289429529462}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:44:59,677] Trial 902 finished with value: 0.7684183121603927 and parameters: {'w1': 0.6953745768730399, 'w2': 0.23674965491472466}. Best is trial 868 with value: 0.7687942045632715.
[I 

[I 2021-01-09 02:45:01,898] Trial 937 finished with value: 0.7682923591705126 and parameters: {'w1': 0.664959627197309, 'w2': -0.02318321667743181}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:01,969] Trial 938 finished with value: 0.7687542290210087 and parameters: {'w1': 0.6828253903013164, 'w2': 0.1332038730357894}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:02,044] Trial 939 finished with value: 0.7686154623147942 and parameters: {'w1': 0.6131868708990151, 'w2': 0.1646911371514712}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:02,115] Trial 940 finished with value: 0.7687658543868761 and parameters: {'w1': 0.6592515251764781, 'w2': 0.05899059783950157}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:02,185] Trial 941 finished with value: 0.7687769566353455 and parameters: {'w1': 0.6133570471941157, 'w2': 0.10595785658838841}. Best is trial 868 with value: 0.7687942045632715.
[I 2021

[I 2021-01-09 02:45:04,309] Trial 976 finished with value: 0.768448385204306 and parameters: {'w1': 0.7395384611100357, 'w2': 0.24430023337934353}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:04,367] Trial 977 finished with value: 0.768793827108103 and parameters: {'w1': 0.6926776291096652, 'w2': 0.0905538641091935}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:04,426] Trial 978 finished with value: 0.7682733364433174 and parameters: {'w1': 0.7421818498469899, 'w2': -0.02794775680623489}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:04,485] Trial 979 finished with value: 0.7686709418914389 and parameters: {'w1': 0.6392277472097041, 'w2': 0.15572291003187022}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-01-09 02:45:04,542] Trial 980 finished with value: 0.768764740767466 and parameters: {'w1': 0.6921155717571956, 'w2': 0.06136391927698994}. Best is trial 868 with value: 0.7687942045632715.
[I 2021-

[I 2021-01-09 02:45:06,653] Trial 1015 finished with value: 0.7687942063365509 and parameters: {'w1': 0.9652028950123748, 'w2': 0.1303357948548996}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:06,712] Trial 1016 finished with value: 0.7687175284723162 and parameters: {'w1': 0.982696014979056, 'w2': 0.2155420656131882}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:06,771] Trial 1017 finished with value: 0.7686007798153867 and parameters: {'w1': 0.9901721137666254, 'w2': 0.27203042996022714}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:06,832] Trial 1018 finished with value: 0.7687630401926022 and parameters: {'w1': 0.9618045775620073, 'w2': 0.18029448552044217}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:06,890] Trial 1019 finished with value: 0.7687916789068749 and parameters: {'w1': 0.9688925956211991, 'w2': 0.14367953326910368}. Best is trial 1015 with value: 0.7687942063365509

[I 2021-01-09 02:45:09,009] Trial 1054 finished with value: 0.7687878316507368 and parameters: {'w1': 0.6441485054669924, 'w2': 0.10110404964091664}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:09,066] Trial 1055 finished with value: 0.7687805796983129 and parameters: {'w1': 0.8710365236547695, 'w2': 0.146613658305149}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:09,124] Trial 1056 finished with value: 0.7687018661094239 and parameters: {'w1': 0.9668861941488008, 'w2': 0.22056883203428396}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:09,188] Trial 1057 finished with value: 0.7687007917554841 and parameters: {'w1': 0.9299320906729007, 'w2': 0.05107054831767343}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:09,250] Trial 1058 finished with value: 0.7686706984455214 and parameters: {'w1': 0.392685555437719, 'w2': 0.09570873842684555}. Best is trial 1015 with value: 0.7687942063365509

[I 2021-01-09 02:45:11,309] Trial 1093 finished with value: 0.7687649431746336 and parameters: {'w1': 0.9589697246903466, 'w2': 0.17815491773089198}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:11,383] Trial 1094 finished with value: 0.7687530961488517 and parameters: {'w1': 0.5999663666785415, 'w2': 0.11756361819366401}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:11,451] Trial 1095 finished with value: 0.23124427433124894 and parameters: {'w1': -0.8203640518163937, 'w2': -0.06749079985056593}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:11,518] Trial 1096 finished with value: 0.7687447764289541 and parameters: {'w1': 0.8299609322261473, 'w2': 0.06267575768594705}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:11,578] Trial 1097 finished with value: 0.7687812099724467 and parameters: {'w1': 0.9156764014913418, 'w2': 0.15338878202655498}. Best is trial 1015 with value: 0.7687942063

[I 2021-01-09 02:45:13,748] Trial 1132 finished with value: 0.7685919078456449 and parameters: {'w1': 0.6557067187628587, 'w2': 0.18248582391954504}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:13,807] Trial 1133 finished with value: 0.7687761401668836 and parameters: {'w1': 0.76229911842042, 'w2': 0.07483752737002725}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:13,866] Trial 1134 finished with value: 0.768482170988455 and parameters: {'w1': 0.8847799871683604, 'w2': -0.0029557078282938243}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:13,925] Trial 1135 finished with value: 0.7687924056980687 and parameters: {'w1': 0.9360491617103252, 'w2': 0.11604069698495062}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:13,990] Trial 1136 finished with value: 0.768715178877223 and parameters: {'w1': 0.7208258714031849, 'w2': 0.15899087458791147}. Best is trial 1015 with value: 0.76879420633655

[I 2021-01-09 02:45:16,094] Trial 1171 finished with value: 0.7686119993536163 and parameters: {'w1': 0.8383780375510459, 'w2': 0.2263570118749128}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:16,153] Trial 1172 finished with value: 0.768763506058411 and parameters: {'w1': 0.9024466293143333, 'w2': 0.16867963522887008}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:16,211] Trial 1173 finished with value: 0.7687583202297154 and parameters: {'w1': 0.9974579695177275, 'w2': 0.08361431810437545}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:16,271] Trial 1174 finished with value: 0.7687735169801245 and parameters: {'w1': 0.7831795427043337, 'w2': 0.13843562562482037}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:16,332] Trial 1175 finished with value: 0.7685453443173991 and parameters: {'w1': 0.6297517685739451, 'w2': 0.006156732410518326}. Best is trial 1015 with value: 0.76879420633655

[I 2021-01-09 02:45:18,418] Trial 1210 finished with value: 0.7687937756830028 and parameters: {'w1': 0.9605533009795836, 'w2': 0.12582509056227403}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:18,478] Trial 1211 finished with value: 0.7687134575296588 and parameters: {'w1': 0.7448955823865532, 'w2': 0.04485645482472202}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:18,537] Trial 1212 finished with value: 0.768695606940091 and parameters: {'w1': 0.8833163311550926, 'w2': 0.2043589326045291}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:18,596] Trial 1213 finished with value: 0.7687879719931285 and parameters: {'w1': 0.8322596563176354, 'w2': 0.09415718621471358}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:18,655] Trial 1214 finished with value: 0.7687367171277902 and parameters: {'w1': 0.6554143156708112, 'w2': 0.13592207618206903}. Best is trial 1015 with value: 0.768794206336550

[I 2021-01-09 02:45:20,738] Trial 1249 finished with value: 0.7685633104234478 and parameters: {'w1': 0.75775007349991, 'w2': 0.2193157839759582}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:20,797] Trial 1250 finished with value: 0.7687674386852819 and parameters: {'w1': 0.9050126522133185, 'w2': 0.16579741807231635}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:20,857] Trial 1251 finished with value: 0.7687905402082287 and parameters: {'w1': 0.7887351116152775, 'w2': 0.11963066883915231}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:20,916] Trial 1252 finished with value: 0.7686333987817459 and parameters: {'w1': 0.9997723456813307, 'w2': 0.03203930743259717}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:20,982] Trial 1253 finished with value: 0.7687899484395886 and parameters: {'w1': 0.6707013004485932, 'w2': 0.07815753663994034}. Best is trial 1015 with value: 0.7687942063365509

[I 2021-01-09 02:45:23,123] Trial 1288 finished with value: 0.7687932087402731 and parameters: {'w1': 0.6278566676254118, 'w2': 0.08055218679407242}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:23,187] Trial 1289 finished with value: 0.7687559450487349 and parameters: {'w1': 0.8304719183324331, 'w2': 0.16081032361420947}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:23,247] Trial 1290 finished with value: 0.7643895307386805 and parameters: {'w1': 0.15768997365458493, 'w2': 0.2248833089441437}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:23,306] Trial 1291 finished with value: 0.7686632620720486 and parameters: {'w1': 0.9114029873938964, 'w2': 0.0376361700488419}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:23,365] Trial 1292 finished with value: 0.7687926970225276 and parameters: {'w1': 0.7094035550153114, 'w2': 0.10299300588009676}. Best is trial 1015 with value: 0.76879420633655

[I 2021-01-09 02:45:25,499] Trial 1327 finished with value: 0.7687927033556681 and parameters: {'w1': 0.9298832102074878, 'w2': 0.13548087183247304}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:25,558] Trial 1328 finished with value: 0.7687911086709122 and parameters: {'w1': 0.8957257037792551, 'w2': 0.1066765645839497}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:25,620] Trial 1329 finished with value: 0.7687843803425375 and parameters: {'w1': 0.9988271504486435, 'w2': 0.16256263582247846}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:25,679] Trial 1330 finished with value: 0.7686964153021333 and parameters: {'w1': 0.8437877944506726, 'w2': 0.0449225987222609}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:25,739] Trial 1331 finished with value: 0.7687768978638025 and parameters: {'w1': 0.8366727419616591, 'w2': 0.08268825187610912}. Best is trial 1015 with value: 0.768794206336550

[I 2021-01-09 02:45:27,848] Trial 1366 finished with value: 0.7685827886301018 and parameters: {'w1': 0.7463435416050053, 'w2': 0.2103877139441906}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:27,908] Trial 1367 finished with value: 0.7686460597427336 and parameters: {'w1': 0.8621268629436386, 'w2': 0.030815592400219236}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:27,977] Trial 1368 finished with value: 0.7677270896564495 and parameters: {'w1': 0.800421621950244, 'w2': -0.07829121063003461}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:28,037] Trial 1369 finished with value: 0.7687899578126364 and parameters: {'w1': 0.898309850770179, 'w2': 0.1047324894528406}. Best is trial 1015 with value: 0.7687942063365509.
[I 2021-01-09 02:45:28,097] Trial 1370 finished with value: 0.7687525464322638 and parameters: {'w1': 0.8429377446683256, 'w2': 0.16552317433260771}. Best is trial 1015 with value: 0.768794206336550

[I 2021-01-09 02:45:30,198] Trial 1405 finished with value: 0.7687805462593316 and parameters: {'w1': 0.962599298092851, 'w2': 0.0989587431919994}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:30,257] Trial 1406 finished with value: 0.768758961903502 and parameters: {'w1': 0.9126346074741248, 'w2': 0.17452441056283047}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:30,317] Trial 1407 finished with value: 0.7687933523758976 and parameters: {'w1': 0.9666830060732635, 'w2': 0.12447138856780185}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:30,377] Trial 1408 finished with value: 0.7686240244673067 and parameters: {'w1': 0.9968303032842023, 'w2': 0.2639887975710006}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:30,436] Trial 1409 finished with value: 0.7687832059249793 and parameters: {'w1': 0.9009416659233592, 'w2': 0.09559739995652122}. Best is trial 1387 with value: 0.768794222802716.
[I 2

[I 2021-01-09 02:45:32,564] Trial 1444 finished with value: 0.7687747086238245 and parameters: {'w1': 0.8925383237518834, 'w2': 0.08646419359052766}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:32,622] Trial 1445 finished with value: 0.766973034799842 and parameters: {'w1': 0.9989292148576815, 'w2': 0.7181039286162388}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:32,685] Trial 1446 finished with value: 0.7678264943750442 and parameters: {'w1': 0.9963369107498462, 'w2': 0.5048876095035143}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:32,745] Trial 1447 finished with value: 0.7687320990018012 and parameters: {'w1': 0.9611748529218352, 'w2': 0.20223568973469763}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:32,807] Trial 1448 finished with value: 0.7687937898692373 and parameters: {'w1': 0.868954904322609, 'w2': 0.11453830050965645}. Best is trial 1387 with value: 0.768794222802716.
[I 2

[I 2021-01-09 02:45:34,947] Trial 1483 finished with value: 0.7687498665005668 and parameters: {'w1': 0.8960006593625788, 'w2': 0.17742768522824676}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:35,006] Trial 1484 finished with value: 0.7687922417963948 and parameters: {'w1': 0.799328370523883, 'w2': 0.11769425511826317}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:35,066] Trial 1485 finished with value: 0.7687245124062378 and parameters: {'w1': 0.9597061785132368, 'w2': 0.06253235370826701}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:35,126] Trial 1486 finished with value: 0.768793739964091 and parameters: {'w1': 0.9970931057714949, 'w2': 0.14045249075283275}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:35,185] Trial 1487 finished with value: 0.7687103157853626 and parameters: {'w1': 0.9995954881975777, 'w2': 0.22326319305710485}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:45:37,318] Trial 1522 finished with value: 0.7687174083959738 and parameters: {'w1': 0.8683007521059303, 'w2': 0.19051937373690575}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:37,379] Trial 1523 finished with value: 0.7687549758249262 and parameters: {'w1': 0.9221696745369605, 'w2': 0.07544390512108198}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:37,438] Trial 1524 finished with value: 0.7687790250390044 and parameters: {'w1': 0.7833774265851171, 'w2': 0.13338893014029393}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:37,498] Trial 1525 finished with value: 0.7687782643021779 and parameters: {'w1': 0.9583414280562055, 'w2': 0.16397070273279502}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:37,558] Trial 1526 finished with value: 0.7687916606674305 and parameters: {'w1': 0.8453513494184425, 'w2': 0.10250530797384191}. Best is trial 1387 with value: 0.768794222802716.


[I 2021-01-09 02:45:39,703] Trial 1561 finished with value: 0.7687912406535582 and parameters: {'w1': 0.9991245612936901, 'w2': 0.119483501757998}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:39,765] Trial 1562 finished with value: 0.7686855597928144 and parameters: {'w1': 0.823302276830378, 'w2': 0.04043276889500927}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:39,828] Trial 1563 finished with value: 0.7687715868922522 and parameters: {'w1': 0.8841639285666457, 'w2': 0.08330588905162484}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:39,891] Trial 1564 finished with value: 0.7687873477988094 and parameters: {'w1': 0.9996647194134674, 'w2': 0.15795633164978928}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:39,955] Trial 1565 finished with value: 0.7686673530274298 and parameters: {'w1': 0.7469198924342438, 'w2': 0.18325972153282624}. Best is trial 1387 with value: 0.768794222802716.
[I 

[I 2021-01-09 02:45:42,130] Trial 1600 finished with value: 0.7687651473550805 and parameters: {'w1': 0.7873735633114778, 'w2': 0.07002960900645835}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:42,189] Trial 1601 finished with value: 0.7687859086559818 and parameters: {'w1': 0.9647429672472714, 'w2': 0.15476392454640953}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:42,248] Trial 1602 finished with value: 0.7687817194102617 and parameters: {'w1': 0.9177233232390705, 'w2': 0.09552593860162782}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:42,307] Trial 1603 finished with value: 0.7687791562616737 and parameters: {'w1': 0.7352112098048545, 'w2': 0.1250430650570775}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:42,367] Trial 1604 finished with value: 0.7686547695840806 and parameters: {'w1': 0.8731650557333756, 'w2': 0.0335855213796698}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:45:44,516] Trial 1639 finished with value: 0.7687929936668245 and parameters: {'w1': 0.9983331374785667, 'w2': 0.1262038260375795}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:44,579] Trial 1640 finished with value: 0.7687484134248304 and parameters: {'w1': 0.9990346923967625, 'w2': 0.0776811170294199}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:44,640] Trial 1641 finished with value: 0.7678029652384567 and parameters: {'w1': 0.849245725049296, 'w2': -0.07723030626664795}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:44,700] Trial 1642 finished with value: 0.7687738640362191 and parameters: {'w1': 0.7940836475172134, 'w2': 0.13994239470819128}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:44,760] Trial 1643 finished with value: 0.768697884590709 and parameters: {'w1': 0.9029402717077352, 'w2': 0.20792206111207112}. Best is trial 1387 with value: 0.768794222802716.
[I 

[I 2021-01-09 02:45:46,933] Trial 1678 finished with value: 0.7687692887222592 and parameters: {'w1': 0.9992432784182307, 'w2': 0.09211073290030072}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:47,002] Trial 1679 finished with value: 0.7686346598366649 and parameters: {'w1': 0.7504519541390733, 'w2': 0.19526926513076398}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:47,063] Trial 1680 finished with value: 0.7686036347950848 and parameters: {'w1': 0.9139210219829097, 'w2': 0.021775000649576604}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:47,124] Trial 1681 finished with value: 0.7685367522724635 and parameters: {'w1': 0.8613121185187702, 'w2': 0.25787179331278864}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:47,185] Trial 1682 finished with value: 0.7687938759999469 and parameters: {'w1': 0.9603862588651407, 'w2': 0.12805114618382474}. Best is trial 1387 with value: 0.768794222802716.

[I 2021-01-09 02:45:49,349] Trial 1717 finished with value: 0.7687824624142947 and parameters: {'w1': 0.9115951921903469, 'w2': 0.1510378260283339}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:49,410] Trial 1718 finished with value: 0.7685089196403028 and parameters: {'w1': 0.7011831394504466, 'w2': 0.0013933733024587003}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:49,471] Trial 1719 finished with value: 0.7687729832230504 and parameters: {'w1': 0.9980921644871655, 'w2': 0.09512893677688547}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:49,531] Trial 1720 finished with value: 0.7655677043557597 and parameters: {'w1': 0.7522983756733043, 'w2': 0.8120706508581169}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:49,593] Trial 1721 finished with value: 0.7686793588883734 and parameters: {'w1': 0.8610089979355318, 'w2': 0.206217348121474}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:45:51,766] Trial 1756 finished with value: 0.7687926053186546 and parameters: {'w1': 0.7675078760884021, 'w2': 0.09545072713875964}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:51,829] Trial 1757 finished with value: 0.7687509487076003 and parameters: {'w1': 0.7030487182588941, 'w2': 0.13874241663924552}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:51,891] Trial 1758 finished with value: 0.7687318900081674 and parameters: {'w1': 0.9247269702684151, 'w2': 0.06359743914874975}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:51,963] Trial 1759 finished with value: 0.7687122618327487 and parameters: {'w1': 0.9662468752745145, 'w2': 0.2147458893929206}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:52,028] Trial 1760 finished with value: 0.768793984929962 and parameters: {'w1': 0.7768697645267125, 'w2': 0.10640664948319332}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:45:54,200] Trial 1795 finished with value: 0.7687159370807931 and parameters: {'w1': 0.6601406784937769, 'w2': 0.040454570079289556}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:54,262] Trial 1796 finished with value: 0.768502326841133 and parameters: {'w1': 0.7304777703087961, 'w2': 0.0004749996219496373}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:54,324] Trial 1797 finished with value: 0.7687747304098276 and parameters: {'w1': 0.7030677070572965, 'w2': 0.06818628314905581}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:54,385] Trial 1798 finished with value: 0.7687905845402115 and parameters: {'w1': 0.6489528864788298, 'w2': 0.09831583463871288}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:54,446] Trial 1799 finished with value: 0.7687572737415935 and parameters: {'w1': 0.7412084276244492, 'w2': 0.06163947967737673}. Best is trial 1387 with value: 0.768794222802716

[I 2021-01-09 02:45:56,631] Trial 1834 finished with value: 0.3134310987784861 and parameters: {'w1': -0.20845110020774224, 'w2': 0.1557978881303941}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:56,692] Trial 1835 finished with value: 0.7687795101575601 and parameters: {'w1': 0.6676333709105513, 'w2': 0.11321049511585665}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:56,753] Trial 1836 finished with value: 0.7686710698208754 and parameters: {'w1': 0.7374375387551302, 'w2': 0.03241186448753296}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:56,815] Trial 1837 finished with value: 0.7687836710308112 and parameters: {'w1': 0.8112974530113342, 'w2': 0.08643440034506852}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:56,878] Trial 1838 finished with value: 0.768766407396697 and parameters: {'w1': 0.727243036201539, 'w2': 0.13392815675748848}. Best is trial 1387 with value: 0.768794222802716.
[

[I 2021-01-09 02:45:59,086] Trial 1873 finished with value: 0.7687883410885517 and parameters: {'w1': 0.8077971185162854, 'w2': 0.12608960655159124}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:59,148] Trial 1874 finished with value: 0.76872833052993 and parameters: {'w1': 0.6548267691092625, 'w2': 0.043817161073282776}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:59,211] Trial 1875 finished with value: 0.7686595521183949 and parameters: {'w1': 0.3999564469787861, 'w2': 0.09959163052208268}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:59,273] Trial 1876 finished with value: 0.7686812638970095 and parameters: {'w1': 0.7583716599588015, 'w2': 0.18094002485940927}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:45:59,334] Trial 1877 finished with value: 0.7686812849230358 and parameters: {'w1': 0.5690844273146979, 'w2': 0.1357633293377827}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:46:01,535] Trial 1912 finished with value: 0.7687063150139011 and parameters: {'w1': 0.6875202554870351, 'w2': 0.15515487393444283}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:01,598] Trial 1913 finished with value: 0.7687916626940354 and parameters: {'w1': 0.9644916515724345, 'w2': 0.11698020012564997}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:01,661] Trial 1914 finished with value: 0.7687085642920501 and parameters: {'w1': 0.8052816313461689, 'w2': 0.1805774840208384}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:01,723] Trial 1915 finished with value: 0.768692027195804 and parameters: {'w1': 0.8986356121750533, 'w2': 0.04635009217670094}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:01,786] Trial 1916 finished with value: 0.7687937951890753 and parameters: {'w1': 0.6272801729325924, 'w2': 0.08202025217280422}. Best is trial 1387 with value: 0.768794222802716.
[I

[I 2021-01-09 02:46:03,988] Trial 1951 finished with value: 0.7687924667495423 and parameters: {'w1': 0.9651541426971141, 'w2': 0.11976531863522744}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:04,051] Trial 1952 finished with value: 0.765425806544925 and parameters: {'w1': 0.8838982202447101, 'w2': 0.9885366650648874}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:04,113] Trial 1953 finished with value: 0.768784376036002 and parameters: {'w1': 0.9354448359758526, 'w2': 0.15224032747803118}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:04,177] Trial 1954 finished with value: 0.7686770543852398 and parameters: {'w1': 0.8660932012098065, 'w2': 0.20842414498921302}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:04,240] Trial 1955 finished with value: 0.7687803869175187 and parameters: {'w1': 0.9090261359860009, 'w2': 0.0933509255275164}. Best is trial 1387 with value: 0.768794222802716.
[I 2

[I 2021-01-09 02:46:06,453] Trial 1990 finished with value: 0.7687837883205717 and parameters: {'w1': 0.8573812627033924, 'w2': 0.09142976608792144}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:06,516] Trial 1991 finished with value: 0.7687917820104009 and parameters: {'w1': 0.9340449296988401, 'w2': 0.1383305598960673}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:06,577] Trial 1992 finished with value: 0.7687085554256534 and parameters: {'w1': 0.9984885028581579, 'w2': 0.058022226540416104}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:06,640] Trial 1993 finished with value: 0.7687937341376018 and parameters: {'w1': 0.8467266081838606, 'w2': 0.11845776978573022}. Best is trial 1387 with value: 0.768794222802716.
[I 2021-01-09 02:46:06,702] Trial 1994 finished with value: 0.7687517710025515 and parameters: {'w1': 0.8918832313321212, 'w2': 0.17563848211687}. Best is trial 1387 with value: 0.768794222802716.
[I 

In [32]:
print('The best roc_auc_score for the study is: ',best_study_score)

The best roc_auc_score for the study is:  0.768794222802716


In [33]:
print(('The best study parameters for the classifier are: ',best_study_params))

('The best study parameters for the classifier are: ', {'w1': 0.9675154320896099, 'w2': 0.1304126470684431})


### Training the Tuned Weighted Aggregating Classifier on the whole Training set & making the predictions on the Test set.

In [34]:
# Training the voting classifier on whole Traing set
voting_clf.fit(X_train.values, y_train)

VotingClassifier(estimators=[('logistic_Reg',
                              LogisticRegression(C=0.0031160262723627184,
                                                 class_weight={0: 1.0, 1: 14},
                                                 l1_ratio=0.4046164083668398,
                                                 n_jobs=5, penalty='elasticnet',
                                                 random_state=42,
                                                 solver='saga')),
                             ('Random_Forest',
                              RandomForestClassifier(class_weight={0: 1.0,
                                                                   1: 13},
                                                     max_depth=18,
                                                     min_samples_leaf=0.0010028580411287713,
                                                     n_estimators=570, n_jobs=5,
                                                     ra...
             

In [35]:
# Making Predictions on the Test Set using voting classifier
y_preds_voting = voting_clf.predict_proba(X_test.values)[:,1]

In [36]:
# Importing best Neural Net Classifier i.e. Dense Neural Net with RELU activation and Nadam optimizer
neural_clf = keras.models.load_model('Best_model_Elu_eq_Learn_relu.h5')

In [39]:
# Making Predictions on the Test Set using Neural classifier
y_preds_neural = neural_clf.predict_proba(X_test)

In [43]:
# combining the predictions from the Voting  & Neural classifier with the study's best votes.
y_combined_final = (y_preds_voting.reshape(-1,1)) * best_study_params['w1'] + y_preds_neural * best_study_params['w2']

In [43]:
print('The training set roc_auc score of the weighted Aggregating classifier is: ', best_study_score)

The training set roc_auc score of the weighted Aggregating classifier is:  0.768794222802716


In [44]:
# Computing the the final roc_auc score of the weighted Aggregating classifier on the test set.
print('The final test set roc_auc score of the combined voting & neural classifier using the best weights is: ',
     roc_auc_score(y_test,y_combined_final))

The final test set roc_auc score of the combined voting & neural classifier using the best weights is:  0.7726740445628326


In [45]:
# Saving the 5 fold Training set Predictions dataframe df_pred as a csv file.
df_pred.to_csv('Voting_Neural_Folds_Preds.csv',index=False)

### Calculating R_R ratio of the tuned Weighted Aggregating classifier using study's best weights.

In [46]:
# Calculating the length of each test fold
len_fold = len(df_pred)//5

In [47]:
len_fold

46126

In [48]:
best_study_params['w1']

0.9675154320896099

In [49]:
best_study_params['w2']

0.1304126470684431

In [55]:
# Calculating roc_auc score for each of the 5 folds and storing it to a list roc_auc_agg
roc_auc_agg = []
for i in range(5):
    if i == 4:
        y_pred_combined = (df_pred.iloc[len_fold*i:,0]* best_study_params['w1'] + 
                           df_pred.iloc[len_fold*i:,1]* best_study_params['w2'])
        roc_auc_agg.append(roc_auc_score(df_pred.iloc[len_fold*i:,2],y_pred_combined))
        
        
        
    else: 
        y_pred_combined = (df_pred.iloc[len_fold*i:len_fold*(i+1),0]* best_study_params['w1'] + 
                           df_pred.iloc[len_fold*i:len_fold*(i+1),1]* best_study_params['w2'])
        
        roc_auc_agg.append(roc_auc_score(df_pred.iloc[len_fold*i:len_fold*(i+1),2],y_pred_combined)) 

In [56]:
print('The reward associated with the Aggregating classifier using roc_auc metric is: ',np.mean(roc_auc_agg))

The reward associated with the Aggregating classifier using roc_auc metric is:  0.7689233205883108


In [57]:
print('The risk associated with the Aggregating Classifier using roc_auc metric is: ',np.std(roc_auc_agg))

The risk associated with the Aggregating Classifier using roc_auc metric is:  0.004576671065945797


In [58]:
R_R_Ratio_Aggregator = np.mean(roc_auc_agg)/np.std(roc_auc_agg)

In [59]:
print('The reward risk ratio for the Aggregating Classifier using roc_auc metric is: ',R_R_Ratio_Aggregator)

The reward risk ratio for the Aggregating Classifier using roc_auc metric is:  168.00930403535745


In [60]:
print('5 fold CV roc_auc scores for the tuned Aggregating Classifier are: ',roc_auc_agg)

5 fold CV roc_auc scores for the tuned Aggregating Classifier are:  [0.763284109827825, 0.764284166520123, 0.7753680619254809, 0.7719635459532233, 0.7697167187149022]


In [63]:
# Saving the test set predictions of the Weighted Aggregating classifier as a dataframe
df_weighted_test_preds = pd.DataFrame(y_combined_final.flatten(),columns=['Weighted_Test_Preds'])

In [64]:
# Saving the test set predictions of the Weighted Aggregating classifier as the csv file
df_weighted_test_preds.to_csv('Weighted_Classifier_Test_Preds.csv',index=False)

#### R_R Ratio for the Weighted Aggregating classifier  using roc_auc metric is:  168.00930403535745

## Observations:
### 1)  The Weighted Aggregating Classifier has fitted the Training set very well, with no signs of overfitting. Also its test set roc_auc score is the highest of any classifier tested till now on this dataset.
### 2) Its R_R ratio is also on the higher side of the spectrum. Thus this classifier is one of the contenders for the best classifier on this dataset.



## Conclusion: Considering all the aspects such as overfitting, test set roc_auc score, R_R ratio etc. , Weighted Aggregating Classifier is one of the best choices. Hence we would use it for making forecasts on future observations and in calculating the optimal threshold probability for this dataset.